github-trigger-workflows-and-wait/src/index.ts

```diff
@@ -0,0 +1,134 @@
+      const repoList = await Promise.all(unprocessedRepos.map(async (repoRecord) => {
+        const splitedRepo = repoRecord.repo.split("/");
+        if (splitedRepo.length != 2) {
+          throw new Error("🔴 Invalid repos format for: " + repoRecord.repo);
+        }
+        const owner = splitedRepo[0]
+        const repo = splitedRepo[1]
+
+        const response = await octokit.rest.actions.listRepoWorkflows({owner,repo});
+        const workflowMatchigTheName = response.data.workflows.filter(workflow=>workflow.name === workflowName)
+        if(!workflowMatchigTheName.length){
+          throw new Error(`🔴 No workflow found for repository ${owner}/${repo}`);
+        }
+        const workflow_id = workflowMatchigTheName[0].id;
+        return { owner, repo, workflow_id }
+      }));
+
+      core.info(`✅ Successfully parsed repo list: ${JSON.stringify(repoList)}`);
+      return repoList;
+    }
+
+    async function triggerRepoWorkflows(repos:ParsedRepo[]) {
+      let success = true
+      core.info("⏳⏳⏳ Triggering workflows list: ");
+      await Promise.all(repos.map(async ({owner,repo, workflow_id})=>{
+        core.info(`Triggering workflow for :${owner}/${repo} ...`);
+
+        const response = await octokit.rest.actions.createWorkflowDispatch({owner ,repo ,workflow_id, ref:"main", inputs: {environment}});
+        if (response.status !== 204) {
+          core.error(`Failed to trigger workflow dispatch for :${owner}/${repo}`);
+          success = false;
+        }
+        else {
+          core.info(`Successfully trigerred workflow dispatch for :${owner}/${repo}`);
+        }
+      }));
+
+      if(success) {
+        core.info("✅ Successfully triggered all workflows")
+      }
+      else {
+        throw new Error('🔴 There were failures triggering some of the workflow dispatches. Look for above workflow dispatch failures');
+      }
+      return success;
+    };
+
+    async function waitForWorkflowStatuses(repos:ParsedRepo[]) {
+      let oneWorkflowFailed = false;
+      let attemptNumber = 1;
+      const maxAttempts = Math.ceil(waitTimeout/checkInterval);
+      const remainingWorkflowsMap = new Map(repos.map(repo => [`${repo.owner}/${repo.repo}`, true]));
+
+      core.info('⏳⏳⏳ Waiting for workflows to report status ...');
+      while(remainingWorkflowsMap.size > 0 && oneWorkflowFailed === false && attemptNumber<=maxAttempts) {
+        await sleep(checkInterval);
+        await Promise.all(repos.map(async ({owner, repo, workflow_id}) => {
+          // Skip checking for workflow status check if it already reported successfull
+          const noSuccessReportYet = remainingWorkflowsMap.get(`${owner}/${repo}`);
+          if(!noSuccessReportYet) {
+            attemptNumber+=1;
+            return;
+          }
+          const response = await octokit.rest.actions.listWorkflowRuns({owner, repo, workflow_id, per_page:10, created: `>${workflowStartISOTimestamp}`});
```

Comment:
Instead of using start timestamp, is there an execution id or similar that is returned in response of `createWorkflowDispatch`?

Reply:
No unfortunately not, Last time i have looked the response was Github is planning on adding the functionality in 2025.
But it looks like they have changed their mind last week and the functionality may be there sooner:
https://github.com/orgs/community/discussions/9752